fendermint/vm/topdown/src/convert.rs

```diff
@@ -0,0 +1,159 @@
+        if value.block_hash.len() != 32 {
+            return Err(anyhow!("invalid block hash length, expecting 32"));
+        }
+
+        let mut array = [0u8; 32];
+        array.copy_from_slice(&value.block_hash[0..32]);
+
+        Ok(Self {
+            height: U256::from(value.height),
+            block_hash: array,
+        })
+    }
+}
+
+impl From<gateway_getter_facet::ParentFinality> for IPCParentFinality {
+    fn from(value: gateway_getter_facet::ParentFinality) -> Self {
+        IPCParentFinality {
+            height: value.height.as_u64(),
+            block_hash: value.block_hash.to_vec(),
+        }
+    }
+}
+
+/// Converts a Rust type FVM address into its underlying payload
+/// so it can be represented internally in a Solidity contract.
+fn addr_payload_to_bytes(payload: Payload) -> Bytes {
+    match payload {
+        Payload::Secp256k1(v) => ethers::types::Bytes::from(v),
+        Payload::Delegated(d) => {
+            let addr = d.subaddress();
+            let b = ethers::abi::encode(&[Token::Tuple(vec![
+                Token::Uint(U256::from(d.namespace())),
+                Token::Uint(U256::from(addr.len())),
+                Token::Bytes(addr.to_vec()),
+            ])]);
+            ethers::types::Bytes::from(b)
+        }
+        _ => unimplemented!("unexpected payload type"),
+    }
+}
+
+fn convert_addr(addr: Address) -> gateway_router_facet::FvmAddress {
+    gateway_router_facet::FvmAddress {
+        addr_type: addr.protocol() as u8,
+        payload: addr_payload_to_bytes(addr.into_payload()),
+    }
+}
+
+pub fn encode_commit_parent_finality_call(
+    finality: IPCParentFinality,
+    validator_set: ValidatorSet,
+) -> anyhow::Result<Vec<u8>> {
+    let commit_function = gateway_router_facet::GATEWAYROUTERFACET_ABI
+        .functions
+        .get(COMMIT_PARENT_FINALITY_FUNC_NAME)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi function map does not have {}",
+                COMMIT_PARENT_FINALITY_FUNC_NAME
+            )
+        })?
+        .get(0)
+        .ok_or_else(|| {
+            anyhow!(
+                "report bug, abi vec does not have {}",
+                COMMIT_PARENT_FINALITY_FUNC_NAME
+            )
+        })?;
+
+    let validators = validator_set.validators.unwrap_or_default();
+
+    let mut addresses = vec![];
+    let mut weights = vec![];
+    for validator in validators {
+        let raw_address = validator.worker_addr.unwrap_or(validator.addr);
+        let addr = Address::from_str(&raw_address)?;
+        addresses.push(convert_addr(addr));
+        weights.push(U256::from_dec_str(&validator.weight)?);
+    }
+
+    let data = ethers::contract::encode_function_data(
+        commit_function,
+        gateway_router_facet::CommitParentFinalityCall {
+            finality: gateway_router_facet::ParentFinality::try_from(finality)?,
+            validators: addresses,
+            weights,
+        },
```

Comment:
@adlrocha what's the outcome of the discussion about the configuration number? Is there an issue we can link here to add it to the contract?